File: solutions/csharp/rail-fence-cipher/1/RailFenceCipher.cs
public readonly record struct RailFenceCipher(int Rails)
{
    private int Period => 2 * (Rails - 1);

    public string Encode(string input) =>
        string.Concat(ZigZagIndexes(input.Length)
            .Select(index => input[index]));

    public string Decode(string input)
    {
        var decoded = new char[input.Length];

        foreach (var (targetIndex, encodedIndex) in DecodingIndexes(input.Length))
        {
            decoded[targetIndex] = input[encodedIndex];
        }

        return new string(decoded);
    }

    private IEnumerable<int> ZigZagIndexes(int length) =>
        Enumerable.Range(0, length).OrderBy(TargetRail);

    private IEnumerable<(int TargetIndex, int EncodedIndex)> DecodingIndexes(int length) =>
        ZigZagIndexes(length).Select((targetIndex, encodedIndex) => (targetIndex, encodedIndex));

    private int TargetRail(int index)
    {
        var position = index % Period;

        return position < Rails
            ? position
            : Period - position;
    }
}

File: solutions/java/rail-fence-cipher/1/src/main/java/RailFenceCipher.java
import java.util.Comparator;
import java.util.stream.IntStream;
import java.util.stream.Collectors;

class RailFenceCipher {
    private final int rows;

    RailFenceCipher(int rows) {
        this.rows = rows;
    }

    String getEncryptedData(String message) {
        return zigZagIndexes(message.length())
            .mapToObj(index -> String.valueOf(message.charAt(index)))
            .collect(Collectors.joining());
    }

    String getDecryptedData(String message) {
        var decoded = new char[message.length()];
        var encodedIndex = 0;

        for (var targetIndex : zigZagIndexes(message.length()).toArray()) {
            decoded[targetIndex] = message.charAt(encodedIndex);
            encodedIndex++;
        }

        return new String(decoded);
    }

    private IntStream zigZagIndexes(int length) {
        return IntStream.range(0, length)
            .boxed()
            .sorted(Comparator.comparingInt(this::targetRail))
            .mapToInt(Integer::intValue);
    }

    private int targetRail(int index) {
        var period = 2 * (rows - 1);
        var position = index % period;

        return position < rows ? position : period - position;
    }
}
